metric/metric_recorder.go

```diff
@@ -0,0 +1,53 @@
+package metric
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/goadesign/goa"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var reqLabels = []string{"entity", "action"}
+
+// Recorder record metrics
+func Recorder() goa.Middleware {
+	reqCnt := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "wit_requests_total",
```

Comment:
Shouldn't this be the normal http_requests_total just with 2 additional labels (ctrl, action)?

Reply:
Discussed with @aslakknutsen and decided that we will use existing 'http_requests_total' metric with existing labels only.  

Labels (ctrl, action) will NOT be introduced rather we will map (ctrl, action) to existing metrics label as below.
ctrl = handler, action = method

So for space entity, metric would be recorded wrt action would be: 
create 		- http_requests_total{code="200",handler="space",method="post"} 8
delete 		- http_requests_total{code="200",handler="space",method="delete"} 8
show/list 	- http_requests_total{code="200",handler="space",method="get"} 8

Note, show/list are mapped to 'get'.

Reply:
@aslakknutsen some more finding on using http_requests_total metric.

**1) instrumenting http requests using prometheus metrics:**
prometheus has provided http metrics to instrument any http request directly.  I have written example of how to instrument any http handler with prometheus metrics.

https://github.com/nurali-techie/play-go-prometheus/tree/master/httpold-metrics

Here only one-liner will instrument any http handler
http.Handle("/myapp", prometheus.InstrumentHandler("myapp", http.HandlerFunc(myAppHander)))

If we want to do similar thing then we need to figure out how goa and prometheus can be integrated.

**2) prometheus http metrics are deprecated:**
Prometheus has deprecated http request metrics.
See Go Doc: https://godoc.org/github.com/prometheus/client_golang/prometheus#Handler

We are using prometheus.Handler() which is deprecated.
We should use instead promhttp.Handler().  But latest version 0.8.0 has no http metrics.  Latest master branch has two metrics - promhttp_metric_handler_requests_total, promhttp_metric_handler_requests_in_flight (https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/http.go)

**3) prometheus http summary metrics not useful**
If you check deprecation reason in Go Doc:
https://godoc.org/github.com/prometheus/client_golang/prometheus#InstrumentHandler

One imp issue is:
- It uses Summaries rather than Histograms. Summaries are not useful if aggregation across multiple instances is required.

In our case, we will be having multiple instance of OSIO (here instance is multiple machine having diff IP addr).  Due to limitation with summary metrics the overall picture across all instance cannot be interpreted.

**4) other open source project reference**
I have found few OS projects which have their own metrics for tracking request count sort of thing.
For ex, https://github.com/coreos/etcd/blob/master/proxy/httpproxy/metrics.go

It seems that it is better to have our own metrics names because if we go with same metrics name like prometheus 'http_requests_total' then we also need to align definition of prometheus metrics. In future, if prometheus change their definition then we also need to adapt the changes.

Overall, we need to consider above points and need to re-visit whether we want to go with prometheus metrics or define our own.